core/common/validation/statebased/validator_keylevel.go

```diff
@@ -268,7 +268,12 @@ func (klv *KeyLevelValidator) Validate(cc string, blockNum, txNum uint64, rwsetB
 	}
 
 	// return the decision of the policy evaluator
-	return policyEvaluator.Evaluate(blockNum, txNum, rwset.NsRwSets, cc, signatureSet)
+	err := policyEvaluator.Evaluate(blockNum, txNum, rwset.NsRwSets, cc, signatureSet)
+	if err != nil {
+		// if endorsement policy check fails, log the endorser identities
+		logger.Warnw("Endorsment policy failure", "error", err, "chaincode", cc, "endorsingIdentities", protoutil.LogMessageForSerializedIdentities(signatureSet))
```

Comment:
is it worth logging the policy as well?

Reply:
yes, good idea, the ccEP is a byte array, is there an existing function to turn that into a human readable policy?

Reply:
yeah that is tricky since we support different types of policy. We could try to unmarshal the bytes as 

https://github.com/hyperledger/fabric/blob/main/core/policy/application.go#L148 and print with %+v or %#v

Reply:
Thanks @ale-linux I tried it out and that approach works well for both signature policies and channel config reference policies, I'll push another one...